File: php/wordy/wordy.php
<?php

const OPERATOR_REGEX = '/(plus|minus|multiplied|divided)/';
const NUM_REGEX      = '/(-?\d)+/';

function calculate($string) {
    preg_match_all(OPERATOR_REGEX, $string, $operators);

    if (! $operators[0]) {
        throw new InvalidArgumentException;
    }

    preg_match_all(NUM_REGEX, $string, $nums);

    if (count($nums[0]) != count($operators[0]) + 1) {
        throw new InvalidArgumentException;
    }

    for ($i = 1, $j = 0; $i < count($nums[0]); ++ $i, ++ $j) {
        $nums[0][$i] = $operators[0][$j]($nums[0][$i - 1], $nums[0][$i]);
    }

    return $nums[0][count($nums[0]) - 1];
}

function plus($a, $b) {
    return $a + $b;
}

function minus($a, $b) {
    return $a - $b;
}

function multiplied($a, $b) {
    return $a * $b;
}

function divided($a, $b) {
    return $a / $b;
}

File: python/wordy/wordy.py
import re

OPERATOR_REGEX = '(plus|minus|multiplied|divided)'

def calculate(sentence):
    operators = re.findall(OPERATOR_REGEX, sentence)
    if operators == None:
        raise ValueError

    nums = map(int, re.findall(r"\-?\d+", sentence))
    if len(nums) != len(operators) + 1:
        raise ValueError

    i = 1
    j = 0
    while i < len(nums):
        if operators[j] == 'plus':
            nums[i] = nums[i - 1] + nums[i]
        if operators[j] == 'minus':
            nums[i] = nums[i - 1] - nums[i]
        if operators[j] == 'multiplied':
            nums[i] = nums[i - 1] * nums[i]
        if operators[j] == 'divided':
            nums[i] = nums[i - 1] / nums[i]
        i += 1
        j += 1
    return nums[len(nums) - 1]
